File: heap-sort.rb
class HeapSort

  def initialize(array)
    @target = array.length - 1
    @array = array
  end

  # traverse subarray A[0..n/2] from bottom-up and run max_heapify on each one
  def build_heap
    counter = @target/2
    while counter >= 0
      max_heapify(counter)
      counter -= 1
    end
    @array
  end

  # make root value greater than the child value
  def max_heapify(parent_index)

    left_child_index = 2 * parent_index + 1
    right_child_index = 2 * parent_index + 2
    largest = parent_index

    # left_child_index <= @target is checked to discard already sorted items
    if left_child_index <= @target && @array[left_child_index] > @array[parent_index]
      largest = left_child_index
    end

    if right_child_index <= @target && @array[right_child_index] > @array[largest]
      largest = right_child_index
    end

    if largest != parent_index
      exchange(parent_index, largest)
      max_heapify(largest);
    end

    @array
  end

  def exchange(first_index, second_index)
    temp = @array[first_index]
    @array[first_index] = @array[second_index]
    @array[second_index] = temp
  end

  def sort
    build_heap
    while @target >= 0
      exchange(0, @target)
      @target -= 1
      max_heapify(0)
    end
    @array
  end
end

array = [7,3,1444,2,1,78,3]
heap_array = HeapSort.new(array)
p heap_array.sort


=begin
Time complexity: O(n lg n)
* call to build_heap takes o(n) time
* each of the n-1 calls to max_heapify takes lg n time
=end

File: heap-sort.py
class HeapSort:

  def __init__(self, array):
    self.target = len(array) - 1
    self.array = array

  def build_heap(self):
    counter = self.target/2

    for i in range(counter, -1, -1):
      self.max_heapify(i)

    return self.array

  # make root value greater than the child value
  def max_heapify(self, parent_index):

    left_child_index = 2 * parent_index + 1
    right_child_index = 2 * parent_index + 2
    largest = parent_index

    if (left_child_index <= self.target and
        self.array[left_child_index] > self.array[parent_index]):
      largest = left_child_index

    if (right_child_index <= self.target and
        self.array[right_child_index] > self.array[largest]):
      largest = right_child_index

    if largest != parent_index:
      self.exchange(parent_index, largest)
      self.max_heapify(largest);

    return self.array


  def exchange(self, first_index, second_index):
    temp = self.array[first_index]
    self.array[first_index] = self.array[second_index]
    self.array[second_index] = temp


  def sort(self):
    self.build_heap()

    while self.target >= 0:
      self.exchange(0, self.target)
      self.target -= 1
      self.max_heapify(0)

    return self.array


array = [6,5,3,1,8,7,2,4]
heap_array = HeapSort(array)
print heap_array.sort()




